File: src/main/java/play/pluv/login/application/RegisterUpdater.java
package play.pluv.login.application;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;
import play.pluv.login.domain.SocialLoginId;
import play.pluv.login.domain.SocialLoginIdRepository;
import play.pluv.member.application.NickNameGenerator;
import play.pluv.member.domain.Member;
import play.pluv.member.domain.NickName;
import play.pluv.member.domain.repository.MemberRepository;
import play.pluv.oauth.domain.OAuthMemberInfo;

@Component
@RequiredArgsConstructor
@Transactional
public class RegisterUpdater {

  private final SocialLoginIdRepository socialLoginIdRepository;
  private final NickNameGenerator nickNameGenerator;
  private final MemberRepository memberRepository;

  public Member registerNewMember(final OAuthMemberInfo memberInfo) {
    final NickName nickName = nickNameGenerator.generateNickName();
    final Member member = new Member(nickName);

    final SocialLoginId loginId = SocialLoginId.builder()
        .oauthMemberInfo(memberInfo)
        .member(member)
        .build();
    socialLoginIdRepository.save(loginId);

    return memberRepository.save(member);
  }

  public void addOtherLoginSource(final Long memberId, final OAuthMemberInfo memberInfo) {
    if (socialLoginIdRepository.existsByOauthMemberInfo(memberInfo)) {
      return;
    }
    final Member member = memberRepository.readById(memberId);

    final SocialLoginId loginId = SocialLoginId.builder()
        .oauthMemberInfo(memberInfo)
        .member(member)
        .build();

    socialLoginIdRepository.save(loginId);
  }
}

File: src/main/kotlin/play/pluv/login/application/RegisterUpdater.kt
package play.pluv.login.application

import jakarta.transaction.Transactional
import org.springframework.stereotype.Component
import play.pluv.login.domain.SocialLoginId
import play.pluv.login.domain.SocialLoginIdRepository
import play.pluv.member.application.MemberReader
import play.pluv.member.application.MemberUpdater
import play.pluv.member.application.NickNameGenerator
import play.pluv.member.domain.Member
import play.pluv.oauth.domain.OAuthMemberInfo

@Component
@Transactional
open class RegisterUpdater(
    private val socialLoginIdRepository: SocialLoginIdRepository,
    private val nickNameGenerator: NickNameGenerator,
    private val memberReader: MemberReader,
    private val memberUpdater: MemberUpdater,
) {

    open fun registerNewMember(oAuthMemberInfo: OAuthMemberInfo): Member {
        val nickName = nickNameGenerator.generateNickName()
        val member = memberUpdater.register(Member(nickName))

        val socialLoginId = SocialLoginId(
            member,
            oAuthMemberInfo,
        )

        socialLoginIdRepository.save(socialLoginId)

        return member
    }

    open fun addOtherLoginSource(memberId:Long, oAuthMemberInfo: OAuthMemberInfo) {
        if(socialLoginIdRepository.existsById(memberId)) {
            return
        }
        val member = memberReader.readById(memberId)

        val socialLoginId = SocialLoginId(
            member,
            oAuthMemberInfo,
        )

        socialLoginIdRepository.save(socialLoginId)
    }
}